use-cases/eurac/hpo.py

```diff
@@ -0,0 +1,202 @@
+    print(f"Ray Tune experiment path: {result_grid.experiment_path}")
+
+    # Get the best result based on the last 10 iterations' average
+    best_result = result_grid.get_best_result(
+        scope="last-10-avg",
+        metric=args.metric,
+        mode="min"
+    )
+    print(f"Best result: {best_result}")
+
+    # Print a dataframe with all trial results
+    result_df = result_grid.get_dataframe()
+    print(f"All results dataframe: {result_df}")
+    print(f"All result columns: {result_df.columns}")
+
+    # Plot the results for all trials
+    plot_results(result_grid, metric=args.metric, filename="ray-loss-plot.png")
+    plot_results(result_grid, metric="train_loss",
+                 filename="ray-train_loss-plot.png")
+
+
+def plot_results(result_grid, metric="loss", filename="plot.png"):
+    """
+    Plot the results for all trials and save the plot to a file.
+
+    Args:
+    - result_grid: Results from Ray Tune trials.
+    - metric: The metric to plot (e.g., 'loss').
+    - filename: Name of the file to save the plot.
+    """
+    ax = None
+    for result in result_grid:
+        label = f"lr={result.config['lr']:.6f}, batch size={result.config['batch_size']}"
+        if ax is None:
+            ax = result.metrics_dataframe.plot(
+                "training_iteration", metric, label=label)
+        else:
+            result.metrics_dataframe.plot(
+                "training_iteration", metric, ax=ax, label=label)
+
+    ax.set_title(
+        f"{metric.capitalize()} vs. Training Iteration for All Trials")
+    ax.set_ylabel(metric.capitalize())
+
+    # Save the plot to a file
+    plt.savefig(filename)
+
+    # Show the plot
+    plt.show()
+
+
+# Main entry point for script execution
+if __name__ == "__main__":
+
+    # Parse command-line arguments
+    parser = argparse.ArgumentParser(
+        description='Hyperparameter Optimization with Ray Tune')
+    parser.add_argument(
+        '--load_old_results', type=bool,
+        default=False,
+        help='Set this to true if you want to load results from an older ray run.')
+    parser.add_argument(
+        '--experiment_path', type=str,
+        default='~/ray_results/Eurac-Ray-Experiment',
+        help='Directory where the results of the previous run are stored. \
+        Set this only if load_old_results is set to True. \
+        Defaults to ~/ray_results/Eurac-Ray-Experiment')
+    parser.add_argument(
+        '--num_samples', type=int,
+        default=10, help='Number of trials to run')
+    parser.add_argument(
+        '--ngpus', type=int,
+        help='Number of GPUs available on node.')
+    parser.add_argument(
+        '--ncpus', type=int,
+        help='Number of CPUs available on node.')
+    parser.add_argument(
+        '--metric', type=str, default='loss',
+        help='Metric to optimise.')
+    parser.add_argument(
+        '--max_iterations', type=int,
+        default='20', help='Maximum iterations per trial')
```

Comment:
Make sure to not put multiple arguments on the same line (see earlier comment) and also make sure to put closing parenthesis on new line.